dawn/prototype/AtlasFromNetcdf.cpp

```diff
@@ -0,0 +1,298 @@
+    std::cout << "lat / long variable not found\n";
+    return false;
+  }
+  if(lon.size() != lat.size()) {
+    std::cout << "lat / long not of consistent sizes!\n";
+    return false;
+  }
+
+  int numNodes = lat.size();
+
+  // define nodes and associated properties for Atlas meshs
+  mesh.nodes().resize(numNodes);
+  atlas::mesh::Nodes& nodes = mesh.nodes();
+  auto xy = atlas::array::make_view<double, 2>(nodes.xy());
+  auto lonlat = atlas::array::make_view<double, 2>(nodes.lonlat());
+
+  // we currently don't care about parts, so myPart is always 0 and remotde_idx == glb_idx
+  auto glb_idx_node = atlas::array::make_view<atlas::gidx_t, 1>(nodes.global_index());
+  auto remote_idx = atlas::array::make_indexview<atlas::idx_t, 1>(nodes.remote_index());
+  auto part = atlas::array::make_view<int, 1>(nodes.partition());
+
+  // no ghosts currently (ghost = false always) and no flags are set
+  auto ghost = atlas::array::make_view<int, 1>(nodes.ghost());
+  auto flags = atlas::array::make_view<int, 1>(nodes.flags());
+
+  auto radToLat = [](double rad) { return rad / (0.5 * M_PI) * 90; };
+  auto radToLon = [](double rad) { return rad / (M_PI)*180; };
+
+  for(int nodeIdx = 0; nodeIdx < numNodes; nodeIdx++) {
+    xy(nodeIdx, atlas::LON) = radToLon(lon[nodeIdx]);
+    xy(nodeIdx, atlas::LAT) = radToLat(lat[nodeIdx]);
+
+    lonlat(nodeIdx, atlas::LON) = xy(nodeIdx, atlas::LON);
+    lonlat(nodeIdx, atlas::LAT) = xy(nodeIdx, atlas::LAT);
+
+    glb_idx_node(nodeIdx) = nodeIdx;
+    remote_idx(nodeIdx) = nodeIdx;
+
+    part(nodeIdx) = myPart;
+    ghost(nodeIdx) = false;
+    atlas::mesh::Nodes::Topology::reset(flags(nodeIdx));
+  }
+
+  return true;
+}
+
+static bool CellsFromNetCDF(const netCDF::NcFile& dataFile, atlas::Mesh& mesh) {
+  auto [cellToVertex, vertexPerCell, ncells] = Load2DField<int>(dataFile, "vertex_of_cell");
+  if(vertexPerCell != 3) {
+    std::cout << "not a triangle mesh\n";
+    return false;
+  }
+
+  // define cells and associated properties
+  mesh.cells().add(new atlas::mesh::temporary::Triangle(), ncells);
+  auto cells_part = atlas::array::make_view<int, 1>(mesh.cells().partition());
+  atlas::mesh::HybridElements::Connectivity& node_connectivity = mesh.cells().node_connectivity();
+  atlas::array::ArrayView<atlas::gidx_t, 1> glb_idx_cell =
+      atlas::array::make_view<atlas::gidx_t, 1>(mesh.cells().global_index());
+
+  for(int cellIdx = 0; cellIdx < ncells; cellIdx++) {
+    // indices in netcdf are 1 based, data is column major
+    atlas::idx_t tri_nodes[3] = {cellToVertex[0 * ncells + cellIdx] - 1,
+                                 cellToVertex[1 * ncells + cellIdx] - 1,
+                                 cellToVertex[2 * ncells + cellIdx] - 1};
+    node_connectivity.set(cellIdx, tri_nodes);
+    glb_idx_cell[cellIdx] = cellIdx;
+    cells_part(cellIdx) = myPart;
+  }
+
+  return true;
+}
+
+static bool AddNeighborList(const netCDF::NcFile& dataFile, atlas::Mesh& mesh,
```

Comment:
Remove code duplication by introducing a template (or can it be solved with runtime polymorphism even?)

Reply:
unfortunately runtime polymorphism is not possible because the connectivities use a Interface<Impl> pattern, solved with template.